homeassistant/components/mqtt/client.py

```diff
@@ -544,11 +544,11 @@ def _client_unsubscribe(topic: str) -> int:
             _raise_on_error(result)
             return mid
 
-        if any(other.topic == topic for other in self.subscriptions):
-            # Other subscriptions on topic remaining - don't unsubscribe.
-            return
-
         async with self._paho_lock:
+            if any(other.topic == topic for other in self.subscriptions):
```

Comment:
Ideally we keep these in a dict so we don't have to iterate them every time we unsub in case there are thousands

Reply:
We could add a new function that checks `_simple_subscriptions` and if its not in there iterates the `_wildcard_subscriptions` to see if its in there.

Reply:
I'm working on a change now.

Reply:
Not currently handling wildcards, as they're not THAT important.